### PR TITLE
Add functional update_and_attend KV-cache op + reference cache

### PR DESCRIPTION
**Pull request by kiymetakdemir**
**Summary**
Adds a neutral kvcache::update_and_attend custom op for KV-cache attention. The op is functional to the tracer with the cache held off-graph via a process-global registry, so the exported .pte does not bake in cache size etc. Ships an eager reference cache (ContiguousReferenceCache) with runtime-selectable static/dynamic sizing and a hard capacity bound.

- op_update_and_attend.py (new) — the kvcache::update_and_attend custom op and the off-graph cache registry
  (install_cache/set_active/uninstall_cache).
- op_update_and_attend_reference.py (new) — CacheConfig/AttendSpec and ContiguousReferenceCache (update_and_fetch, static/dynamic sizing, hard capacity) plus the neutral attend SDPA mechanism.
- test_update_and_attend.py (new) — functional-graph + prefill/decode parity tests across static and dynamic sizing.
- targets.bzl (modified) — adds the update_and_attend_py python_library.
- BUCK (modified) — adds the test_update_and_attend python_test.
 
**Test plan**
pytest -v extension/llm/custom_ops/test_update_and_attend.py
Verifies the exported graph is functional (no buffer inputs/mutations, one op call per layer) and that outputs match cacheless causal attention for prefill and incremental decode under both static and dynamic sizing.

### CODE REVIEW

File: extension/llm/custom_ops/op_update_and_attend_reference.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+class MaskKind(Enum):
+    NONE = "none"  # decode: q_len == 1, one row sees all of history
+    CAUSAL = "causal"  # prefill: q_len > 1, row i sees keys 0..offset+i
+
+
+@dataclass
+class AttendSpec:
+    kind: MaskKind
+    offset: int = 0  # cached tokens preceding this step
```
Comment (digantdesai) [inline extension/llm/custom_ops/op_update_and_attend_reference.py]: is this sufficient for different kinds of attn? curious

File: extension/llm/custom_ops/BUCK [hunk 2]
```diff
@@ -91,3 +91,25 @@ fbcode_target(_kind = runtime.python_test,
...
+    srcs = [
+        "test_update_and_attend.py",
+    ],
+    deps = [
+        ":update_and_attend_py",
+        "//caffe2:torch",
+    ],
+)
+
+fbcode_target(_kind = runtime.python_binary,
```
Comment (metascroy) [inline extension/llm/custom_ops/BUCK]: Looks like you need to rebase?

File: extension/llm/custom_ops/op_update_and_attend.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+from __future__ import annotations
+
+from typing import Dict, Optional
+
+import torch
+
+from executorch.extension.llm.custom_ops.op_update_and_attend_reference import attend
+
+_REGISTRY: Dict[str, object] = {}
+_CURRENT_KEY: Optional[str] = None
```
Comment (metascroy) [inline extension/llm/custom_ops/op_update_and_attend.py]: I know this is just a python reference, but maybe wrap this in a CacheRegistry class for neatness?

File: extension/llm/custom_ops/op_update_and_attend.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+def _(
+    q: torch.Tensor,
+    k: torch.Tensor,
+    v: torch.Tensor,
+    position: torch.Tensor,
+    layer_id: int,
+    scale: float,
+    out_dtype: torch.dtype,
+) -> torch.Tensor:
+    return torch.empty_like(q, dtype=out_dtype)
```
Comment (metascroy) [inline extension/llm/custom_ops/op_update_and_attend.py]: Is this the right shape? Shouldn't final dim be based on v? See https://docs.pytorch.org/docs/2.13/generated/torch.nn.functional.scaled_dot_product_attention.html

File: extension/llm/cache/update_and_attend.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    if _CURRENT_KEY is None:
+        raise RuntimeError(
+            "update_and_attend called with no active cache; "
+            "call install_cache(...) then set_active(...)"
+        )
+    return _REGISTRY[_CURRENT_KEY]
+
+
+@torch.library.custom_op("kvcache::update_and_attend", mutates_args=())
+def update_and_attend(
```
Comment (metascroy) [inline extension/llm/cache/update_and_attend.py:65]: Can we have a doc string describing shape of each arg?

For example, the op should natively handle GQA, so make sure q/k/v shapes are described as such

File: extension/llm/custom_ops/op_update_and_attend_reference.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+) -> torch.Tensor:
+    """Eager attend mechanism: apply SDPA to fetched K/V per the mask semantic."""
+    n_q_heads = q.shape[1]
+    n_kv_heads = k.shape[1]
+    if n_q_heads != n_kv_heads:
+        rep = n_q_heads // n_kv_heads
+        k = k.repeat_interleave(rep, dim=1)
+        v = v.repeat_interleave(rep, dim=1)
+
+    qf = q.to(torch.float32)
```
Comment (metascroy) [inline extension/llm/custom_ops/op_update_and_attend_reference.py]: Why not use F.scaled_dot_product_attention?

File: extension/llm/cache/test_update_and_attend.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    def _export(self, seq_len):
+        x = torch.randn(1, seq_len, self.hidden)
+        pos = _positions(0, seq_len)
+        idx = torch.arange(seq_len, dtype=torch.long)
+        return torch.export.export(self.model, (x, pos, idx), strict=True)
+
+    def test_graph_is_functional(self):
+        # Export needs no cache: _export installs none, so the op traces via its
+        # fake kernel only -- the cleanest statement that the cache is off-graph.
+        ep = self._export(seq_len=5)
```
Comment (metascroy) [inline extension/llm/cache/test_update_and_attend.py:137]: Can we run functionalization on the graph?

ep = ep.run_decompositions({})

File: extension/llm/cache/test_update_and_attend.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+            n_kv_heads=self.n_kv_heads,
+            head_dim=self.head_dim,
+            sizing=sizing,
+            capacity=capacity,
+        )
+
+    def _export(self, seq_len):
+        x = torch.randn(1, seq_len, self.hidden)
+        pos = _positions(0, seq_len)
+        idx = torch.arange(seq_len, dtype=torch.long)
```
Comment (metascroy) [inline extension/llm/cache/test_update_and_attend.py:128]: Can we run functionalization on the graph? Sometimes export succeeds where functionalization fails (as an aside, ET always runs functionalization):

```
ep = ep.run_decompositions({})
```

File: extension/llm/custom_ops/test_update_and_attend.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+        ep = self._export(seq_len)
+        for sizing, cap in [
+            (CacheSizing.DYNAMIC, seq_len),
+            (CacheSizing.STATIC, seq_len),
+        ]:
+            with self.subTest(sizing=sizing):
+                cache = ContiguousReferenceCache(self._config(sizing, cap))
+                install_cache(self.cache_key, cache)
+                set_active(self.cache_key)
```
Comment (metascroy) [inline extension/llm/custom_ops/test_update_and_attend.py]: What is difference between set_active and install?

File: extension/llm/cache/test_update_and_attend.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (metascroy) [inline extension/llm/cache/test_update_and_attend.py:1]: Make sure this is run by the unittest job in CI (look at raw logs to confirm)

File: extension/llm/cache/reference_cache.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (metascroy) [inline extension/llm/cache/reference_cache.py:1]: I take it this is just an example implementation? 

At runtime we do not have to use things like AttendSpec for implementation?

File: extension/llm/custom_ops/BUCK (hunk 2 above)
Comment (kiymetakdemir) [reply extension/llm/custom_ops/BUCK]: Fixed this.

File: extension/llm/custom_ops/op_update_and_attend.py (hunk 4 above)
Comment (kiymetakdemir) [reply extension/llm/custom_ops/op_update_and_attend.py]: You are right, attn_weight is [B, H, q_len, k_len] and value is [B, H, k_len, D_v], so attn_weight @ value is [B, H, q_len, D_v].

File: extension/llm/cache/reference_cache.py (hunk 11 above)
Comment (kiymetakdemir) [reply extension/llm/cache/reference_cache.py:1]: Correct, and for the contiguous cases here the mask is always causal so it is not really functional now, I put it to use later for explicit but I can drop it for simplicity.

File: extension/llm/custom_ops/test_update_and_attend.py (hunk 9 above)
Comment (kiymetakdemir) [reply extension/llm/custom_ops/test_update_and_attend.py]: install registers a cache under a key once (at creation), while active selects which installed cache the op uses for a given forward. They look identical with one cache, but distinct with multi-sequence, where we install each once and switch active per step.

File: extension/llm/custom_ops/op_update_and_attend_reference.py (hunk 1 above)
Comment (kiymetakdemir) [reply extension/llm/custom_ops/op_update_and_attend_reference.py]: As is, no, it only covers causal. It's meant to also carry things like a window or a full mask for other cases, so it can grow to cover them; I just kept it minimal since we only have causal now.

**Review by metascroy — APPROVED** [review]

**Comment by metascroy** [comment]
Looks good! Thinking ahead to where c++ stuff will live, maybe it's better to move these files to a dedicated export/llm/cache folder.

Also mark as experimental
